### PR TITLE
[cosmwasm] Osmosis multiple fee token support

### PR DESCRIPTION
**Pull request by 0xfirefist**

### CODE REVIEW

File: target_chains/cosmwasm/contracts/pyth/Cargo.toml [hunk 1]
```diff
@@ -11,10 +11,12 @@ crate-type = ["cdylib", "rlib"]
 [features]
 # IMPORTANT: if you want to build for injective, enable the default feature below
 # default=["injective"]
+default=["osmosis"]
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/Cargo.toml:14]: do you want this to always be the default?

**Review by jayantk — COMMENTED** [review]
I think this is generally fine. I have a couple questions / comments that should be easy to address, but there are enough that I'd like to look one more time before approving.

I have one other question which i will slack you

File: target_chains/cosmwasm/contracts/pyth/Cargo.toml (hunk 1 above)
Comment (0xfirefist) [reply target_chains/cosmwasm/contracts/pyth/Cargo.toml]: oh yeah

File: target_chains/cosmwasm/contracts/pyth/src/contract.rs [hunk 2]
```diff
@@ -156,6 +156,62 @@ pub fn execute(
...
+        info.funds.as_ref(),
+        &get_update_fee(&deps, data)?,
+    ));
+}
+
+// it only checks for fee denoms other than the base denom
+#[cfg(feature = "osmosis")]
+fn is_allowed_tx_fees_denom(deps: &Deps, denom: &String) -> StdResult<bool> {
+    let querier = TxfeesQuerier::new(&deps.querier);
+    match querier.denom_pool_id(denom.to_string()) {
```
[the remark this answers is not in the document]
Comment (0xfirefist) [reply target_chains/cosmwasm/contracts/pyth/src/contract.rs:169]: yes

[the remark this answers is not in the document]
Comment (0xfirefist) [reply target_chains/cosmwasm/contracts/pyth/src/contract.rs:169]: yes

File: target_chains/cosmwasm/contracts/pyth/src/contract.rs [hunk 3]
```diff
@@ -156,24 +157,38 @@ pub fn execute(
...
 fn is_fee_sufficient(deps: &Deps, info: MessageInfo, data: &[Binary]) -> StdResult<bool> {
     use cosmwasm_std::has_coins;
+
     let state = config_read(deps.storage).load()?;
-    return Ok(has_coins(
-        info.funds.as_ref(),
-        &get_update_fee(&deps, data)?,
-    ));
+
+    // for any chain other than injective there is only one base denom
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/src/contract.rs:166]: you mean osmosis?

File: target_chains/cosmwasm/contracts/pyth/src/contract.rs [hunk 4]
```diff
@@ -182,6 +197,14 @@ fn is_allowed_tx_fees_denom(deps: &Deps, denom: &String) -> StdResult<bool> {
 fn is_fee_sufficient(deps: &Deps, info: MessageInfo, data: &[Binary]) -> StdResult<bool> {
     let state = config_read(deps.storage).load()?;
 
+    // as with chains other than osmosis
+    // if the base fee amount is set to 0
+    // no funds are required
+    // here also the behavior is same
+    if state.fee.amount.u128() == 0 {
+        return Ok(true);
+    }
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/src/contract.rs:206]: why do you need this special case for 0 here and above? it seems like the logic should work without it

File: target_chains/cosmwasm/contracts/pyth/src/contract.rs [hunk 5]
```diff
@@ -156,24 +157,38 @@ pub fn execute(
...
 }
 
 // it only checks for fee denoms other than the base denom
 #[cfg(feature = "osmosis")]
-fn is_allowed_tx_fees_denom(deps: &Deps, denom: &String) -> StdResult<bool> {
+fn is_allowed_tx_fees_denom(deps: &Deps, denom: &String) -> bool {
+    // TxFeesQuerier uses stargate queries which we can't mock as of now.
+    // The capability has not been implemented in `cosmwasm-std` yet.
+    // Hence hacking it this way to be able to right tests.
+    // FIXME
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/src/contract.rs:180]: Use complete sentences in your comments. "Hence, we are hacking it with a feature flag ..."

File: target_chains/cosmwasm/contracts/pyth/src/contract.rs [hunk 6]
```diff
@@ -182,6 +197,14 @@ fn is_allowed_tx_fees_denom(deps: &Deps, denom: &String) -> StdResult<bool> {
 fn is_fee_sufficient(deps: &Deps, info: MessageInfo, data: &[Binary]) -> StdResult<bool> {
     let state = config_read(deps.storage).load()?;
 
+    // as with chains other than osmosis
+    // if the base fee amount is set to 0
+    // no funds are required
+    // here also the behavior is same
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/src/contract.rs:203]: what's with the formatting here? And again, write a complete sentence with proper grammar and capitalization.

File: target_chains/cosmwasm/contracts/pyth/src/wormhole.rs [hunk 7]
```diff
@@ -1,3 +1,8 @@
+// These types are copied from the Wormhole contract. See the links with each type to see the original code
+// The reason to do so was dependency conflict. Wormhole contracts were using a very old version of a dependency
+// which is not compatible with the one used by osmosis-sdk. And since we weren't using anything else from
+// the Wormhole contract the types are moved here.
```
Comment (jayantk) [inline target_chains/cosmwasm/contracts/pyth/src/wormhole.rs:4]: this comment is more along the lines of what the other comments ^ should be like.

**Review by jayantk — APPROVED** [review]
lgtm. i left you some minor comments about your comments

**Pull request merged**